File: app/src/main/java/com/ed/redditapp/ui/main/MainPresenter.java
package com.ed.redditapp.ui.main;

import com.ed.redditapp.lib.api.Post;

public class MainPresenter {
    private final MainActivity view;
    private final MainModel model;

    public MainPresenter(MainActivity view, MainModel model) {
        this.view = view;
        this.model = model;
    }

    public void onActivityLoaded() {
        model.getMainPagePosts((posts) -> {
            view.updatePosts(posts);
            model.getMainPageSubredditIcons(view::updateSubredditIcons);
        });
    }

    public void onPostInfoClicked(Post post) {
        view.gotoSubreddit(post.getSubreddit());
    }

    public void onPostTitleClicked(Post post) {
        view.gotoPostDetail(post);
    }

    public void onThumbnailClicked(Post post) {
        view.gotoImageView(post);
    }
}

File: app/src/main/java/com/ed/redditapp/ui/main/MainPresenter.kt
package com.ed.redditapp.ui.main

import com.ed.redditapp.lib.api.Post

class MainPresenter(val view: MainActivity, val model: MainModel) {
    fun onActivityLoaded() {
        model.getMainPagePosts { posts ->
            view.updatePosts(posts)
            model.getMainPageSubredditIcons(view::updateSubredditIcons)
        }
    }

    fun onPostInfoClicked(post: Post) {
        view.gotoSubreddit(post.subreddit)
    }

    fun onPostTitleClicked(post: Post) {
        view.gotoPostDetail(post)
    }

    fun onThumbnailClicked(post: Post) {
        view.gotoImageView(post)
    }

    fun onNearEndReached(post: Post) {
        model.getNextPosts(post.after) { posts ->
            view.addNextPosts(posts)
        }
    }
}